sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala

```diff
@@ -699,8 +701,42 @@ case class UnionExec(children: Seq[SparkPlan]) extends SparkPlan {
     }
   }
 
+  private lazy val childrenRDDs = children.map(_.execute())
+
+  override def outputPartitioning: Partitioning = {
+    if (conf.getConf(SQLConf.UNION_OUTPUT_PARTITIONING)) {
+      // Commands like `AppendDataExec` have side effects when creating RDDs, so we
+      // cannot call `execute` on them to determine the partitioning.
+      if (children.exists(_.containsPattern(COMMAND))) {
+        return super.outputPartitioning
+      }
+
+      try {
+        val nonEmptyRdds = childrenRDDs.filter(!_.partitions.isEmpty)
+        if (sparkContext.isPartitionerAwareUnion(nonEmptyRdds)) {
```

Comment:
Actually this only covers limited cases like reused shuffles they have the same partitioner. I would like to extend this to SQL cases, i.e., the `outputPartitioning` is same or compatible for Union's children. But I will leave to follow up works.